### PR TITLE
Fixed SQL models and added a test

### LINKED ISSUE #4: Create database models
- [ ] Create a sketch of database models and relationships
- [ ] Create mock objects of database entries according to the model for development and testing

### PR DESCRIPTION
**Pull request by northwestwitch**
### This PR adds | fixes:
Fix #4: 
- Improved sql database table.
- Separated sql_models from pydantic_models (otherwise it's confusing) and put them on different files named accordingly
- Moved code that will be shared by most tests into fixtures (conftest.py file)
- Renamed SQL table `Individuals` to `Samples` 
- Renamed table `Region` to `Interval`
- Added table `Cases`
- Added new table `Tags` -> will contain data like a gene in a certain genome build, or a custom genomic interval of a certain build
- Added new table `interval_tag`, an in-between table to reflect the relationship between `Intervals` and `Tags`. (for instance one genomic interval can have many tags - APC build 38, genome build 38, etc - but also one tag can have many genomic intervals -like genome build 38- --> Many to many relationship. There must be other ways, but this one looks flexible to me..
-  Added test to check that database schema is created when app lauches

Proposed database schema below. I'm sure it can be improved, but it will help a lot starting populating a draft with demo data.

image

**How to prepare for test**:
- Install the repo locally conda env + poetry install

### How to test:
- Launch the app using the command `uvicorn src.chanjo2.main:app`
- Run tests locally and make sure that the database file created is reflecting the schema above

### Expected outcome:
- [x] the app should work as before
- [x] Tests should pass, both locally and in GitHub action

### Review: 
- [x] Code approved by HS
- [x] Tests executed by CR

This [version](https://semver.org/) is a:
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

File: src/chanjo2/main.py [hunk 1]
```diff
@@ -3,49 +3,39 @@
...
-from .endpoints import individuals, regions
+from .dbutil import engine
+from .endpoints import intervals, samples
+from .models.sql_models import Base, Case, Interval, Sample, Tag
 
 LOG = logging.getLogger(__name__)
 coloredlogs.install(level="INFO")
 
 
-class CoverageInterval(BaseModel):
```
Comment (northwestwitch) [inline src/chanjo2/main.py:17]: This is duplicated. The code is only in src.pydantic models now

**Comment by northwestwitch** [comment]
I think I'm ready with this one now!

File: tests/src/chanjo2/test_main.py [hunk 2]
```diff
@@ -1,31 +1,24 @@
...
-
-
-def test_heartbeat():
+def test_heartbeat(client):
     """Test the function that returns a message if server is running"""
     response = client.get("/")
     assert response.status_code == 200
     assert response.json() == {"message": f"Chanjo2 v{__version__} is up and running!"}
+
+
```
Comment (henrikstranneheim) [inline tests/src/chanjo2/test_main.py:13]: FASTAPI also has a nice way to unit test endpoints. We should implement it soon.

**Review by henrikstranneheim — COMMENTED** [review]
Great work! ⭐  I have some architect suggestions, see comments

File: src/chanjo2/endpoints/intervals.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+@router.get("/interval/{sample_id}/", response_model=CoverageInterval)
+def read_interval(
+    *,
+    session: Session = Depends(get_session),
+    sample_id: str,
+    chromosome: str,
+    start: int,
+    end: int,
+):
+    sample = session.exec(select(Individual).filter(sample_id == sample_id)).first()
```
[the remark this answers is not in the document]
Comment (northwestwitch) [reply src/chanjo2/endpoints/intervals.py:20]: I totally agree. I haven't written this part and I didn't touch it from how it was when I started working on the project.

I think this is one of the next steps. In this PR I only wanted to define the SQL database schema So I had to make some adjustments to this file. I haven't even checked what these endpoints do and if they should be modified in the future

File: src/chanjo2/endpoints/intervals.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    )
+
+
+@router.get("/{sample_id}/target_coverage", response_model=List[float])
+def read_target_coverage(
+    *,
+    session: Session = Depends(get_session),
+    sample_id: str,
+):
+    sample = session.exec(select(Individual).filter(sample_id == sample_id)).first()
```
[the remark this answers is not in the document]
Comment (northwestwitch) [reply src/chanjo2/endpoints/intervals.py:44]: Same here. I wasn't focusing on the endpoints on this PR. I basically left them unchanged (code-wise)

File: src/chanjo2/endpoints/intervals.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    if not sample:
+        raise HTTPException(
+            status_code=status.HTTP_404_NOT_FOUND, detail="Sample not found"
+        )
+    coverage_file: D4File = D4File(sample.coverage_file_path)
+    with open(sample.region_file_path, "rb") as default_region_file:
```
[the remark this answers is not in the document]
Comment (northwestwitch) [reply src/chanjo2/endpoints/intervals.py:50]: Agreed. But I'd like to address it in another PR. There is so much to do about the endpoints

File: tests/src/chanjo2/test_main.py [hunk 6]
```diff
@@ -1,31 +1,24 @@
...
-        db.close()
-
-
-app.dependency_overrides[create_db_and_tables] = override_get_db
-client = TestClient(app)
-
-
-def test_heartbeat():
+def test_heartbeat(client):
     """Test the function that returns a message if server is running"""
```
[the remark this answers is not in the document]
Comment (northwestwitch) [reply tests/src/chanjo2/test_main.py:8]: This is actually my code :) I can add it!

File: tests/src/chanjo2/test_main.py (hunk 2 above)
Comment (northwestwitch) [reply tests/src/chanjo2/test_main.py:13]: I know, but this wasn't the purpose of this PR!

File: src/chanjo2/endpoints/intervals.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (henrikstranneheim) [reply src/chanjo2/endpoints/intervals.py:44]: Sounds good! let´s do these changes in more focused PRs 👍

File: tests/conftest.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+
+@pytest.fixture(name="testdb")
+def testdb_fixture():
+    return TEST_DB
+
+
+@pytest.fixture(name="session")
+def session_fixture():
+    engine = create_engine(TEST_DB, connect_args={"check_same_thread": False})
```
[the remark this answers is not in the document]
Comment (northwestwitch) [reply tests/conftest.py:17]: Ah yes! Will fix!

**Comment by northwestwitch** [comment]
Thanks for the review @henrikstranneheim, I really appreciate the effort also given that it was a big and complicated one! O'll address you comments tomorrow!

File: src/chanjo2/endpoints/intervals.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
+from typing import List
```
Comment (northwestwitch) [inline src/chanjo2/endpoints/intervals.py:1]: Please disregard this file. I basically just renamed it. Before it was named regions

File: src/chanjo2/endpoints/samples.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
+from pathlib import Path
```
Comment (northwestwitch) [inline src/chanjo2/endpoints/samples.py:1]: Please disregard this file. I basically just renamed it. Before it was named individuals

**Review by henrikstranneheim — APPROVED** [review]
Nice. 👍

**Pull request merged**